runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,450 @@
+            if bytes_written == 0 {
+                return Err(Error::WriteFailed);
+            }
+
+            total_written += bytes_written;
+        }
+        Ok(())
+    }
+
+    async fn truncate(&self, len: u64) -> Result<(), Error> {
+        // Create an operation to do the truncate
+        let op = opcode::Ftruncate::new(types::Fd(self.fd.as_raw_fd()), len).build();
+
+        // Submit the operation
+        let (sender, receiver) = oneshot::channel();
+        self.io_sender
+            .clone()
+            .send((op, sender))
+            .await
+            .map_err(|_| {
+                Error::BlobTruncateFailed(
+                    self.partition.clone(),
+                    hex(&self.name),
+                    IoError::new(ErrorKind::Other, "failed to send work"),
+                )
+            })?;
+
+        // Wait for the result
+        let return_value = receiver.await.map_err(|_| {
+            Error::BlobTruncateFailed(
+                self.partition.clone(),
+                hex(&self.name),
+                IoError::new(ErrorKind::Other, "failed to read result"),
+            )
+        })?;
+
+        // If the return value is negative, it indicates an error.
+        if return_value < 0 {
+            return Err(Error::BlobTruncateFailed(
+                self.partition.clone(),
+                hex(&self.name),
+                IoError::new(ErrorKind::Other, format!("error code: {}", return_value)),
+            ));
+        }
+        Ok(())
+    }
+
+    async fn sync(&self) -> Result<(), Error> {
+        // Create an operation to do the sync
+        let op = opcode::Fsync::new(types::Fd(self.fd.as_raw_fd())).build();
+
+        // Submit the operation
+        let (sender, receiver) = oneshot::channel();
+        self.io_sender
+            .clone()
+            .send((op, sender))
+            .await
+            .map_err(|_| {
+                Error::BlobSyncFailed(
+                    self.partition.clone(),
+                    hex(&self.name),
+                    IoError::new(ErrorKind::Other, "failed to send work"),
+                )
+            })?;
+
+        // Wait for the result
+        let return_value = receiver.await.map_err(|_| {
+            Error::BlobSyncFailed(
+                self.partition.clone(),
+                hex(&self.name),
+                IoError::new(ErrorKind::Other, "failed to read result"),
+            )
+        })?;
+        // If the return value is negative, it indicates an error.
+        if return_value < 0 {
+            return Err(Error::BlobSyncFailed(
+                self.partition.clone(),
+                hex(&self.name),
+                IoError::new(ErrorKind::Other, format!("error code: {}", return_value)),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Drop all references to self.fd to close that resource.
+    async fn close(self) -> Result<(), Error> {
+        self.sync().await
```

Comment:
It should not be possible for any outstanding items to exist at this point (as `close` consumes `self` and we await all results).

If we change functions to return a waiter instead of waiting explicitly, we should consider waiting for said completions I think.